[cli][client][tests] update to node14-compatible target (drop support for node12)

EndangeredMassa opened the pull request:
Node 12 is EOL and we're starting to reference packages that don't install on node lower than 14, such as nuxt@3 and remix. Let's update to 14 for now.

---

Reimplements: https://github.com/vercel/vercel/pull/7819

EndangeredMassa committed:
update to node14-compatible target

EndangeredMassa committed:
update node versions

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
update tsconfig targets

TooTallNate committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
add module setting

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
validate that fetch urls are absolute

EndangeredMassa committed:
eval-require to avoid bad transpilation

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
remove eager absolute url check

EndangeredMassa committed:
remove unnecessary eslint-disable

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
update lib

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

styfle reviewed (approved):
Thanks!

TooTallNate commented:
Unfortunately Jest is doing something different with the dynamic requires that makes the Unit tests not fail here, but with the built CLI I am getting this with `vc build`:

```
> [debug] [2022-05-25T19:00:07.084Z] Failed to import "@vercel/static-build": Error: Cannot find module '@vercel/static-build/package.json'
> [debug] [2022-05-25T19:00:07.085Z] Installing Builders: @vercel/static-build
yarn add v1.22.18
warning ../../../package.json: No license field
info No lockfile found.
[1/4] 🔍  Resolving packages...
...
Error! Importing "@vercel/static-build": Cannot find module '/Users/nrajlich/r/.vercel/builders/node_modules/@vercel/static-build/dist/index'
```

TooTallNate commented:
We probably need to do the `eval('require')` stuff in `import-builders.ts` as well.

TooTallNate commented on packages/next/src/templated-launcher.ts:
```diff
@@ -9,9 +9,9 @@ if (!process.env.NODE_ENV) {
   process.env.NODE_ENV = region === 'dev1' ? 'development' : 'production';
 }
 
+// NOTE: `eval('require')` is necessary to avoid bad transpilation to `__webpack_require__`
```
Are we sure this one is necessary? I don't believe this file is compiled by webpack.

TooTallNate replied:
@ijjk could maybe confirm.

ijjk replied:
Yeah this one shouldn't be touched by `ncc` just `tsc` x-ref: https://github.com/vercel/vercel/blob/0b56caba45c4ab0a74e7ae0a451cda033db2db04/packages/next/build.js#L9

TooTallNate commented on packages/next/src/utils.ts:
```diff
@@ -244,8 +244,8 @@ export async function getRoutesManifest(
-  // eslint-disable-next-line @typescript-eslint/no-var-requires
-  const routesManifest: RoutesManifest = require(pathRoutesManifest);
+  // NOTE: `eval('require')` is necessary to avoid bad transpilation to `__webpack_require__`
+  const routesManifest: RoutesManifest = eval('require')(pathRoutesManifest);
```
I'd suggest `fs.readJSON()` instead.

TooTallNate committed:
[cli] Add `vc build` integration test

Adds a `vc build` integration test to ensure the ncc'd CLI works as
expected to supplement the unit tests from #7869.

TooTallNate committed:
Merge branch 'add/cli-vc-build-integration-test' into endangeredmassa/update-node-14-tsconfig

TooTallNate commented:
I pushed a failing CLI test for `vc build` (from #7876) to this branch to make sure that the test is passing before we merge this.

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
add require workaround

EndangeredMassa committed:
revert test setup change

EndangeredMassa committed:
requrie workaround

EndangeredMassa committed:
switch require out for fs.readJSON

EndangeredMassa committed:
revert one of the workarounds

TooTallNate commented on packages/next/test/fixtures/00-i18n-support/index.test.js:
```diff
@@ -29,9 +31,16 @@ async function checkForChange(url, initialValue, hardError) {
 
 describe(`${__dirname.split(path.sep).pop()}`, () => {
-  it('should deploy and pass probe checks', async () => {
+  beforeAll(async () => {
```
It's not a bug, it's just that the first test creates the deployment and the other tests perform some assertions on the behavior of the deployment.

This `index.test.js` file is afaik copy/pasted into all the other `next/test/fixtures` directories so if we're making a change here then it should probably apply to all of them.

EndangeredMassa replied:
I would consider this a bug because it relies on test order to execute properly.

I reverted the change here and will make a change in all of these files in a separate PR.

EndangeredMassa commented:
Holding off on merging until we get a new minor release out. Then this can be in a new major release.

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

styfle commented on packages/node-bridge/tsconfig.json:
```diff
@@ -2,11 +2,13 @@
-    "declaration": true
+    "target": "ES2020",
+    "declaration": true,
+    "module": "ES2020"
```
I'm not sure this is correct. The `module` should likely be `commonjs` since we even changed TS files to JS in this package to preserve CJS

EndangeredMassa replied:
Yeah, not sure how this change made it in. I removed it. [`8edba39` (#7865)](https://github.com/vercel/vercel/pull/7865/commits/8edba39fa5ac28d21ebf43c0cf335e723f241b91)

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
turn @typescript-eslint/no-var-requires rule on

EndangeredMassa committed:
turn rule back off, for now

EndangeredMassa commented on packages/next/src/templated-launcher.ts:
```diff
@@ -10,7 +10,6 @@ if (!process.env.NODE_ENV) {
 
 // @ts-ignore
-// eslint-disable-next-line @typescript-eslint/no-var-requires
 const page = require(__LAUNCHER_PAGE_PATH__);
```
Ahh, that's why the rule was off: https://github.com/vercel/vercel/runs/6698194648?check_suite_focus=true

I'll mess with that in a separate PR. Turning this rule back off.

EndangeredMassa replied:
There are 247 violations of this rule. We can discuss if we want to enforce it or not. Maybe for now we remove these comments disabling the rule on a per-line level.

EndangeredMassa committed:
feedback

EndangeredMassa committed:
Merge branch 'main' into endangeredmassa/update-node-14-tsconfig

EndangeredMassa committed:
set module to commonjs
